### PR TITLE
Added ICount.

### PR DESCRIPTION
**Pull request by xavrin**

### CODE REVIEW

File: nesc/whip6/api/io/EventCount.nc [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * whip6: Warsaw High-performance IPv6.
+ *
+ * Copyright (c) 2017 University of Warsaw
+ * All rights reserved.
+ *
+ * This file is distributed under the terms in the attached LICENSE
+ * files.
+ */
+
+interface EventCount {
```
Comment (konrad-iwanicki) [inline nesc/whip6/api/io/EventCount.nc]: Parametrize this interface with a type.

File: nesc/whip6/api/io/EventCount.nc [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    command error_t stop();
+    // Read the absolute number of events since the beginning of time.
+    // This means that the value returned will count events before the start
+    // (if the start wasn't the first one).
+    // The recommended way of using this is to look at differences between the
+    // number of events in 2 points in time. There's no oveflow problem with
+    // this aproach.
+    // Reads can be done in arbitrary moment in time (regardless of start and
+    // stop).
+    command error_t read(unsigned long long *value);
```
Comment (konrad-iwanicki) [inline nesc/whip6/api/io/EventCount.nc]: Use a parametrized type here.

File: nesc/whip6/api/io/EventCount.nc [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * Copyright (c) 2017 University of Warsaw
+ * All rights reserved.
+ *
+ * This file is distributed under the terms in the attached LICENSE
+ * files.
+ */
+
+interface EventCount {
+    // Initialize all necessary hardware and start counting events.
```
Comment (konrad-iwanicki) [inline nesc/whip6/api/io/EventCount.nc]: Use JavaDoc/Doxygen-like comments.

File: nesc/whip6/api/io/EventCountConfig.nc [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+ * files.
+ */
+
+typedef enum {
+    EVENT_COUNT_MODE_RISING_EDGE,
+    EVENT_COUNT_MODE_FALLING_EDGE,
+    EVENT_COUNT_MODE_BOTH,
+} event_count_mode_t;
+
+interface EventCountConfig {
```
Comment (konrad-iwanicki) [inline nesc/whip6/api/io/EventCountConfig.nc:17]: Is this interface really so generic that it should not have ICount in its name?

File: nesc/whip6/apps/ICountRead/ICountReadPrv.nc [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+ * Copyright (c) 2017 University of Warsaw
+ * All rights reserved.
+ *
+ * This file is distributed under the terms in the attached LICENSE
+ * files.
+ */
+
+#define INTERVAL_1S 1024
+
+module ICountReadPrv {
```
Comment (konrad-iwanicki) [inline nesc/whip6/apps/ICountRead/ICountReadPrv.nc:13]: ICountReadAppMainPrv

File: nesc/whip6/api/io/EventCount.nc (hunk 2 above)
Comment (konrad-iwanicki) [reply nesc/whip6/api/io/EventCount.nc]: Moreover, use a type with a known width, for example, uint64_t.

File: nesc/whip6/apps/ICountRead/ICountReadPrv.nc [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    uses interface Boot;
+    uses interface Timer<TMilli, uint32_t>;
+    uses interface EventCount as ICount;
+}
+
+implementation {
+    int timestamp;
```
Comment (konrad-iwanicki) [inline nesc/whip6/apps/ICountRead/ICountReadPrv.nc]: Never use a type with an unknown width.

File: nesc/whip6/apps/ICountRead/ICountReadPrv.nc [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+    event void Boot.booted() {
+        call ICount.start();
+        timestamp = 0;
+        call Timer.startWithTimeoutFromNow(INTERVAL_1S);
+    }
+
+    event void Timer.fired() {
+        unsigned long long ticks = 0;
```
Comment (konrad-iwanicki) [inline nesc/whip6/apps/ICountRead/ICountReadPrv.nc]: uint64_t

File: nesc/whip6/apps/ICountRead/ICountReadPrv.nc [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        error_t res = SUCCESS;
+
+        timestamp += 1;
+        call Timer.startWithTimeoutFromLastTrigger(INTERVAL_1S);
+
+        res = call ICount.read(&ticks);
+        if (res != SUCCESS)
+            printf("Failed to read ICount\n");
+        else
+            printf("ICount - number of ticks: %u, timestamp: %d\n",
```
Comment (konrad-iwanicki) [inline nesc/whip6/apps/ICountRead/ICountReadPrv.nc]: %lu %llu

File: nesc/whip6/platforms/parts/mcu/cc26xx/timers/HalEventCountPrv.nc [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+
+    uses interface ExternalEvent as ChannelAInterrupt @atmostonce();
+    uses interface ExternalEvent as ChannelBInterrupt @atmostonce();
+}
+
+implementation {
+    uint32_t running;  // mask of TIMER_A and TIMER_B
+    // It counts events that occurred during previous timer runs. It is updated
+    // only when timer overrun happens.
+    unsigned long long evCnt[2] = {0};
```
Comment (konrad-iwanicki) [inline nesc/whip6/platforms/parts/mcu/cc26xx/timers/HalEventCountPrv.nc]: uint64_t

File: nesc/whip6/platforms/parts/mcu/cc26xx/timers/HalEventCountPrv.nc [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+    // Value of timer used when the timer is stopped. Otherwise the timer
+    // register should be read.
+    unsigned cachedTimer[2] = {0};
```
Comment (konrad-iwanicki) [inline nesc/whip6/platforms/parts/mcu/cc26xx/timers/HalEventCountPrv.nc]: uint32_t

**Review by konrad-iwanicki — CHANGES_REQUESTED** [review]
Never use types with unknown widths.

File: nesc/whip6/api/io/EventCountConfig.nc (hunk 4 above)
Comment (xavrin) [reply nesc/whip6/api/io/EventCountConfig.nc:17]: These EventCount/EventCountConfig interfaces together with HalEventCountPub can be used to count events on any pin. I felt it's generic enough to separate it from ICount.

File: nesc/whip6/apps/ICountRead/ICountReadPrv.nc (hunk 5 above)
Comment (xavrin) [reply nesc/whip6/apps/ICountRead/ICountReadPrv.nc:13]: Hm, I followed naming convention used in all apps - suffix "App" for configuration and "Prv" for implementation.
Why should I add "AppMain" here?

File: nesc/whip6/apps/ICountRead/ICountReadPrv.nc (hunk 8 above)
Comment (xavrin) [reply nesc/whip6/apps/ICountRead/ICountReadPrv.nc]: Currently printf implementation doesn't support these flags (%lu and %llu). See:
nesc/whip6/lib/extern/tfp_printf/cortex-m3/tfp_printf.c

**Comment by xavrin** [comment]
I changed logic a bit to polling the counter value.
I had constant problems with dealing with overflow while using the interrupt approach.
Now it's a bit simpler and its correctness is obvious.

**Pull request merged**